galaxy_ng/app/api/v3/viewsets/collection.py

```diff
@@ -107,23 +114,27 @@ def list(self, request, *args, **kwargs):
         serializer = CollectionVersionListSerializer(queryset, many=True, context=context)
         return Response(serializer.data)
 
-    # Custom retrive so we can use the class serializer_class
-    # galaxy_ng.app.api.v3.serializers.CollectionVersionSerializer
-    # which is responsible for building the 'download_url'. The default pulp one doesn't
-    # include the distro base_path which we need (https://pulp.plan.io/issues/6510)
-    # so override this so we can use a different serializer
-    def retrieve(self, request, *args, **kwargs):
-        """
-        Returns a CollectionVersion object.
-        """
-        instance = self.get_object()
-        artifact = ContentArtifact.objects.get(content=instance)
 
-        context = self.get_serializer_context()
-        context["content_artifact"] = artifact
+class UnpaginatedCollectionVersionViewSet(api_base.LocalSettingsMixin,
+                                          ViewNamespaceSerializerContextMixin,
+                                          CollectionListMixin,
+                                          pulp_ansible_views.UnpaginatedCollectionVersionViewSet):
+    pagination_class = None
+    serializer_class = CollectionVersionSerializer
```

Comment:
@fao89 do we need a custom serializer here? I am asking because the default is returning 114000 records with only 3 collections see; https://gist.github.com/rochacbruno/031d624d0356e57fa0d0cb3f6b54b978

Reply:
it is the right serializer, it exposes all versions from all collections, and I could see `vmware.vmware_rest` has 68 versions.
Beyond that, we are exposing `MANIFEST.json` and `FILES.json` which made the response bigger.
But I'm seeing manifest field repeating data from other fields, we may need to work on this, FYI @bmbouter @daviddavis

Reply:
I don't think it's reasonable to include the `MANIFEST.json` and `FILES.json` as part of this API endpoint. That's just way too much data for even a medium sized deployment like hub on cloud.redhat.com.

`FILES.json` is especially problematic. Collections don't have a ignore file and it's very possible that someone could accidentally include their `node_modules` directory in a collection which would add something like 100,000 files for a single collection version.

Reply:
@fao89 @bmbouter are manifest and files required for sync?

Reply:
@rochacbruno tells me that `files` was added for the `ansible-galaxy` client. Can @jborean93, @fao89 and @bmbouter  let me know what this is being used for?

Reply:
I believe we can remove the `files` and `manifest` data from this metadata API. The sync pipeline populates that data during sync for items it needs to freshly fetch here: https://github.com/pulp/pulp_ansible/blob/master/pulp_ansible/app/tasks/collections.py#L842-L844

Can someone open a bugfix request for that against pulp_ansible?

I believe these are to enable a CLI search command at some point to help solve the problem of "where did my content go?" @jborean93 would need to give more details on the plan and timeline tho.

Reply:
it was an old request https://pulp.plan.io/issues/7572 
some background here: https://github.com/pulp/pulp_ansible/pull/505#issuecomment-771903712
@newswangerd you have really valid points, and I can see it was not well designed, maybe we need to move it to another endpoint as docs-blob case, or make a parametrizable endpoint as @rochacbruno suggested

Reply:
I calculated that exposing the metadata without files would return about 150mb of data with 100,000 collection versions, which isn't unreasonable with proper caching and streamed API responses, so I'm fine with keeping the metadata in, if it's something that will improve the client.

The request in pulp/pulp_ansible#505 (comment) seems reasonable, if we can isolate the exact set of collection information that the client needs, rather than just returning everything in metadata and files.

Reply:
Overall I am in favor of removing it if the only thing using it would be pulp's sync code. Pulp's sync code doesn't even use it. We need input from @jborean93 on if the CLI needs this in the unpaginated metadata response or not.

Reply:
I believe for the CLI needs, we can keep it only on retrieve, and have another serializer for unpaginated list

Reply:
There's no plans or timelines in place to add the `provides` command to the `ansible-galaxy` cli. Maybe someone will work on it in the 2.12 development cycle but nothing has been planned or talked about yet. One of the problems with adding a new command that relies on new APIs is that it will only work for v3 endpoints leaving galaxy.ansible.com out in the cold. This might make a sales proposition for AAP/AH but IMO there is little desire to implement a feature that will not work for galaxy.

As for the API itself and what data returns there are 2 issues that seem to be at play here

* Getting the version list for a collection can take some time, as time goes on there will be more versions added requiring even more API calls
    * This certainly doesn't need the collection content, just a collection with their versions. I'm not sure if we enumerate the deps for all versions or just once we select the candidate
* Getting collection content information for a `provides` like command
    * This certainly requires information about the collections plugins.
    * Doesn't necessarily have to be the `FILES.json` just a list of plugins, their type, and maybe description

I'm no API expert so I don't know of an efficient way of exposing this information but something must be out there if tools like `dnf` can do this.